---

# NGÀY 1: Setup Documents Table + Update Chunks

**Thời gian:** 4 giờ  
**Mục tiêu:** Create documents table (70 records) + Update chunk metadata

## Step 1: Extract Metadata từ 70 Files

**Thời gian:** 30 phút  
**Input:** 70 files trong data/raw/  
**Output:** documents_metadata.json

In [1]:
# Run extraction script - navigate to project root
import os
import sys

# Hardcode project root path
project_root = "/home/sakana/Code/RAG-bidding"

print(f"Changing to project root: {project_root}")
os.chdir(project_root)
print(f"Current directory: {os.getcwd()}\n")

# Run the script
!python scripts/migration/002_extract_simple_metadata.py

Changing to project root: /home/sakana/Code/RAG-bidding
Current directory: /home/sakana/Code/RAG-bidding

🔍 Scanning data/raw folders...

✅ Found 68 documents

📁 Câu hỏi thi: 5 documents
   - EXAM-NHCH_30925_bo_sung: NHCH 30.9.25 bo sung theo TB1952 qldt (1)
   - EXAM-Ngân-hàng-câu-hỏi-CC: Ngân hàng câu hỏi CCDT đợt 2
   ... and 3 more

📁 Hồ sơ mời thầu: 45 documents
   - FORM-15-Phu-luc: 15. Phu luc
   - FORM-01-Phụ-lục: 01. Phụ lục
   ... and 43 more

📁 Luật chính: 4 documents
   - LUA-90-2025-QH15: Luat so 90 2025-qh15
   - LUA-57-2024-QH15: Luat so 57 2024 QH15
   ... and 2 more

📁 Mẫu báo cáo: 10 documents
   - TEMPLATE-BC-03B: 03B. Mẫu BCTĐ danh sách đáp ứng về kỹ thuật
   - TEMPLATE-BC-03C: 03C. Mẫu BCTĐ KQLCNT
   ... and 8 more

📁 Nghị định: 1 documents
   - ND-214-4.8.-CP: ND 214 - 4.8.2025 - Thay thế NĐ24

📁 Quyết định: 1 documents
   - QD-1667-BYT: Quyết định-1667-QĐ-BYT

📁 Thông tư: 2 documents
   - TT-0-Lời-văn-thông-tư: 0. Lời văn thông tư
   - TT-00-Quyết-định-Thông-tư: 

**Expected Output:**
```
✅ Found 70 documents

📁 Hồ sơ mời thầu: 46 documents
📁 Mẫu báo cáo: 10 documents
📁 Câu hỏi thi: 6 documents
📁 Luật chính: 4 documents
📁 Thông tư: 2 documents
📁 Nghị định: 1 documents
📁 Quyết định: 1 documents

💾 Saved to: data/processed/documents_metadata.json
```

**Verify:**

In [2]:
import json
from pathlib import Path

# Load metadata (now we're in project root)
metadata_file = Path("data/processed/documents_metadata.json")
with open(metadata_file, 'r', encoding='utf-8') as f:
    documents = json.load(f)

print(f"Total documents: {len(documents)}")

# Group by category
by_category = {}
for doc in documents:
    cat = doc['category']
    if cat not in by_category:
        by_category[cat] = []
    by_category[cat].append(doc)

print("\nBy category:")
for cat, docs in sorted(by_category.items(), key=lambda x: len(x[1]), reverse=True):
    print(f"  {cat}: {len(docs)} documents")

# Show sample
print("\nSample documents:")
for doc in documents[:5]:
    print(f"  {doc['document_id']}: {doc['document_name'][:50]}...")

Total documents: 68

By category:
  Hồ sơ mời thầu: 45 documents
  Mẫu báo cáo: 10 documents
  Câu hỏi thi: 5 documents
  Luật chính: 4 documents
  Thông tư: 2 documents
  Nghị định: 1 documents
  Quyết định: 1 documents

Sample documents:
  LUA-90-2025-QH15: Luat so 90 2025-qh15...
  LUA-57-2024-QH15: Luat so 57 2024 QH15...
  LUA-HỢP-NHẤT-126-2025-về: HỢP NHẤT 126 2025 về Luật đấu thầu...
  LUA-Luat-dau-thau-2023: Luat dau thau 2023...
  ND-214-4.8.-CP: ND 214 - 4.8.2025 - Thay thế NĐ24...


## Step 2: Create Documents Table & Populate

**Thời gian:** 1 giờ  
**Input:** documents_metadata.json  
**Output:** documents table với 70 records

### 🔍 Pre-Migration: Visualize Current Database State

Kiểm tra database trước khi chạy migration:

### 📝 Step 2A: Load Metadata

Load documents từ metadata file:

In [3]:
# Load documents metadata
metadata_file = Path("data/processed/documents_metadata.json")

with open(metadata_file, 'r', encoding='utf-8') as f:
    documents_to_insert = json.load(f)

print(f"📂 Loaded {len(documents_to_insert)} documents from metadata file")
print(f"\n📋 Sample document structure:")
print(json.dumps(documents_to_insert[0], indent=2, ensure_ascii=False))

📂 Loaded 68 documents from metadata file

📋 Sample document structure:
{
  "document_id": "LUA-90-2025-QH15",
  "document_name": "Luat so 90 2025-qh15",
  "category": "Luật chính",
  "document_type": "law",
  "source_file": "data/raw/Luat chinh/Luat so 90 2025-qh15.docx",
  "file_name": "Luat so 90 2025-qh15.docx",
  "total_chunks": 0,
  "status": "active"
}


### 📝 Step 2B: Create Documents Table (Split SQL)

Execute SQL schema từng statement một để tránh lỗi asyncpg:

In [4]:
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text
import os

print("📊 Creating documents table...\n")

# Get DATABASE_URL if not already defined
try:
    DATABASE_URL
except NameError:
    DATABASE_URL = os.getenv(
        "DATABASE_URL_ASYNC",
        "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
    )

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

# Read SQL schema file
schema_file = Path("scripts/migration/001_simple_documents_schema.sql")

with open(schema_file, "r") as f:
    sql = f.read()

# Improved SQL parsing: handle comments properly
statements = []
current = ""
in_function = False
in_comment = False

for line in sql.split("\n"):
    # Skip pure comment lines at the start
    stripped = line.strip()
    if stripped.startswith("--") and not current.strip():
        continue
    
    # Track if we're inside a function definition with $$
    if "$$" in line:
        in_function = not in_function
    
    # Add line to current statement
    current += line + "\n"
    
    # Split on semicolon only if:
    # 1. NOT inside function body
    # 2. Line actually ends with semicolon
    if line.rstrip().endswith(";") and not in_function:
        # Clean up statement
        stmt = current.strip()
        if stmt and not stmt.startswith("--"):
            statements.append(stmt)
        current = ""

# Add last statement if exists (might not have semicolon)
if current.strip() and not current.strip().startswith("--"):
    statements.append(current.strip())

print(f"📝 Found {len(statements)} SQL statements to execute")

# Preview statements
print("\n📋 Statement preview:")
for i, stmt in enumerate(statements, 1):
    first_line = stmt.split("\n")[0].strip()
    if first_line.startswith("--"):
        first_line = [l.strip() for l in stmt.split("\n") if l.strip() and not l.strip().startswith("--")][0]
    print(f"   [{i}] {first_line[:70]}...")

print()

# Execute each statement in SEPARATE transactions (idempotent)
async def create_table():
    executed = 0
    skipped = 0
    
    for i, stmt in enumerate(statements, 1):
        try:
            # Show what we're executing
            first_line = stmt.split("\n")[0].strip()
            if first_line.startswith("--"):
                first_line = [l.strip() for l in stmt.split("\n") if l.strip() and not l.strip().startswith("--")][0]
            
            print(f"   [{i}/{len(statements)}] Executing: {first_line[:60]}...")
            
            # IMPORTANT: Each statement in its own transaction
            async with engine.begin() as conn:
                await conn.execute(text(stmt))
            
            executed += 1
            
        except Exception as e:
            error_msg = str(e).lower()
            if "already exists" in error_msg or "duplicate" in error_msg:
                # Ignore "already exists" errors - idempotent
                print(f"      ⏭️  Already exists, skipping...")
                skipped += 1
                continue
            else:
                print(f"   ❌ Error: {e}")
                print(f"      Statement preview: {stmt[:200]}...")
                # Don't raise - continue with other statements
                continue
    
    print(f"\n✅ Executed {executed}/{len(statements)} statements successfully")
    if skipped > 0:
        print(f"⏭️  Skipped {skipped} statements (already exist)")

await create_table()
print("\n✅ Documents table setup complete!")

await engine.dispose()


📊 Creating documents table...

📝 Found 11 SQL statements to execute

📋 Statement preview:
   [1] CREATE TABLE IF NOT EXISTS documents (...
   [2] CREATE INDEX idx_documents_category ON documents(category);...
   [3] CREATE INDEX idx_documents_type ON documents(document_type);...
   [4] CREATE INDEX idx_documents_status ON documents(status);...
   [5] CREATE INDEX idx_documents_source ON documents(source_file);...
   [6] CREATE OR REPLACE FUNCTION update_documents_updated_at()...
   [7] CREATE TRIGGER trigger_documents_updated_at...
   [8] COMMENT ON TABLE documents IS 'Master table for all legal documents re...
   [9] COMMENT ON COLUMN documents.document_id IS 'Unique identifier followin...
   [10] COMMENT ON COLUMN documents.category IS 'One of 7 categories mapping t...
   [11] COMMENT ON COLUMN documents.status IS 'Document status for toggle func...

   [1/11] Executing: CREATE TABLE IF NOT EXISTS documents (...
   [2/11] Executing: CREATE INDEX idx_documents_category ON documents(ca

### 📝 Step 2C: Insert Documents

Insert 68 documents vào table:

In [5]:
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text
import os

print(f"📥 Inserting {len(documents_to_insert)} documents...\n")

# Get DATABASE_URL if not already defined
try:
    DATABASE_URL
except NameError:
    DATABASE_URL = os.getenv(
        "DATABASE_URL_ASYNC",
        "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
    )

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def insert_documents():
    async with AsyncSessionLocal() as session:
        inserted = 0
        skipped = 0
        
        for doc in documents_to_insert:
            try:
                # Check if exists
                result = await session.execute(
                    text("SELECT id FROM documents WHERE document_id = :doc_id"),
                    {"doc_id": doc["document_id"]}
                )
                existing = result.fetchone()
                
                if existing:
                    skipped += 1
                    if skipped <= 5:  # Show first 5
                        print(f"   ⏭️  Skipped (exists): {doc['document_id']}")
                    continue
                
                # Insert new document
                await session.execute(
                    text("""
                        INSERT INTO documents (
                            document_id, document_name, category, document_type,
                            source_file, file_name, total_chunks, status
                        ) VALUES (
                            :document_id, :document_name, :category, :document_type,
                            :source_file, :file_name, :total_chunks, :status
                        )
                    """),
                    {
                        "document_id": doc["document_id"],
                        "document_name": doc["document_name"],
                        "category": doc["category"],
                        "document_type": doc["document_type"],
                        "source_file": doc["source_file"],
                        "file_name": doc["file_name"],
                        "total_chunks": doc.get("total_chunks", 0),
                        "status": doc.get("status", "active"),
                    }
                )
                
                inserted += 1
                
                if inserted % 10 == 0:
                    print(f"   ✅ Inserted {inserted}/{len(documents_to_insert)}...")
            
            except Exception as e:
                print(f"   ❌ Error inserting {doc['document_id']}: {e}")
        
        await session.commit()
        
        print(f"\n{'='*60}")
        print(f"✅ Inserted: {inserted} documents")
        print(f"⏭️  Skipped: {skipped} documents (already exist)")
        print(f"📊 Total: {inserted + skipped} documents in database")
        print(f"{'='*60}")

await insert_documents()
await engine.dispose()


📥 Inserting 68 documents...

   ⏭️  Skipped (exists): LUA-90-2025-QH15
   ⏭️  Skipped (exists): LUA-57-2024-QH15
   ⏭️  Skipped (exists): LUA-HỢP-NHẤT-126-2025-về
   ⏭️  Skipped (exists): LUA-Luat-dau-thau-2023
   ⏭️  Skipped (exists): ND-214-4.8.-CP

✅ Inserted: 0 documents
⏭️  Skipped: 68 documents (already exist)
📊 Total: 68 documents in database


In [6]:
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text
import os

print("🔍 DATABASE PRE-MIGRATION CHECK")
print("="*60)

# Get database URL
DATABASE_URL = os.getenv(
    "DATABASE_URL_ASYNC",
    "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
)

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def check_database():
    async with AsyncSessionLocal() as session:
        # 1. List all tables
        print("\n📊 Tables in Database:")
        result = await session.execute(text("""
            SELECT tablename 
            FROM pg_tables 
            WHERE schemaname = 'public'
            ORDER BY tablename
        """))
        tables = [row[0] for row in result.fetchall()]
        for table in tables:
            print(f"   • {table}")
        
        # 2. Check if documents table exists
        has_documents = 'documents' in tables
        print(f"\n📁 Documents Table: {'✅ EXISTS' if has_documents else '❌ NOT FOUND'}")
        
        if has_documents:
            # Count documents
            result = await session.execute(text("SELECT COUNT(*) FROM documents"))
            doc_count = result.scalar()
            print(f"   Records: {doc_count}")
        
        # 3. Check vector database (langchain_pg_embedding)
        has_vectors = 'langchain_pg_embedding' in tables
        print(f"\n🗄️  Vector Database: {'✅ EXISTS' if has_vectors else '❌ NOT FOUND'}")
        
        if has_vectors:
            # Count chunks
            result = await session.execute(text("""
                SELECT COUNT(*) FROM langchain_pg_embedding
            """))
            chunk_count = result.scalar()
            print(f"   Total chunks: {chunk_count}")
            
            # Count distinct document_ids
            result = await session.execute(text("""
                SELECT COUNT(DISTINCT cmetadata->>'document_id')
                FROM langchain_pg_embedding
            """))
            unique_docs = result.scalar()
            print(f"   Unique document_ids: {unique_docs}")
            
            # Sample current document_ids
            result = await session.execute(text("""
                SELECT DISTINCT cmetadata->>'document_id'
                FROM langchain_pg_embedding
                ORDER BY cmetadata->>'document_id'
                LIMIT 10
            """))
            
            print(f"\n   📋 Sample current document_ids:")
            for row in result.fetchall():
                doc_id = row[0]
                print(f"      {doc_id}")
            
            # Check for source_file metadata
            result = await session.execute(text("""
                SELECT 
                    COUNT(*) FILTER (WHERE cmetadata->>'source_file' IS NOT NULL) as with_source,
                    COUNT(*) as total
                FROM langchain_pg_embedding
            """))
            row = result.fetchone()
            coverage = (row[0] / row[1] * 100) if row[1] > 0 else 0
            status = "✅" if coverage > 90 else "⚠️"
            print(f"\n   {status} source_file metadata: {row[0]}/{row[1]} ({coverage:.1f}%)")

await check_database()
await engine.dispose()

print("\n" + "="*60)
print("✅ Pre-migration check complete!")
print("="*60)

🔍 DATABASE PRE-MIGRATION CHECK

📊 Tables in Database:
   • documents
   • langchain_pg_collection
   • langchain_pg_embedding

📁 Documents Table: ✅ EXISTS
   Records: 59

🗄️  Vector Database: ✅ EXISTS
   Total chunks: 4708
   Unique document_ids: 25

   📋 Sample current document_ids:
      FORM-05-Mẫu-Báo-cáo-đấu-
      FORM-10-Mẫu-số-10A_E-HSM
      FORM-10-Mẫu-số-10B_E-HSM
      FORM-11-Mẫu-số-11A_E-HSM
      FORM-11-Mẫu-số-11B_E-HSM
      FORM-15-Phu-luc
      FORM-7-Mẫu-số-7A-E-HSMT
      FORM-7-Mẫu-số-7B-E-HSMT
      FORM-8-Mẫu-số-8B-E-HSMT
      FORM-9Mẫu-số-9A_E-HSMT_P

   ✅ source_file metadata: 4705/4708 (99.9%)

✅ Pre-migration check complete!
   • documents
   • langchain_pg_collection
   • langchain_pg_embedding

📁 Documents Table: ✅ EXISTS
   Records: 59

🗄️  Vector Database: ✅ EXISTS
   Total chunks: 4708
   Unique document_ids: 25

   📋 Sample current document_ids:
      FORM-05-Mẫu-Báo-cáo-đấu-
      FORM-10-Mẫu-số-10A_E-HSM
      FORM-10-Mẫu-số-10B_E-HSM
      FORM-11-

### ✅ Expected Output (Step 2)

Sau khi chạy các cell trên, bạn sẽ thấy:

**Step 2A - Load Metadata:**
```
📂 Loaded 68 documents from metadata file
```

**Step 2B - Create Table:**
```
📊 Creating documents table...
📝 Found 15 SQL statements to execute

   [1/15] CREATE TABLE IF NOT EXISTS documents (...
   [2/15] CREATE INDEX idx_documents_category ON documents(category...
   ...
✅ Executed 15 statements successfully
✅ Documents table created!
```

**Step 2C - Insert Documents:**
```
📥 Inserting 68 documents...
   ✅ Inserted 10/68...
   ✅ Inserted 20/68...
   ...
✅ Inserted: 68 documents
⏭️  Skipped: 0 documents
📊 Total: 68 documents in database
```

---

**Verify kết quả:**

In [7]:
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text
import os

# Get database URL with password
DATABASE_URL = os.getenv(
    "DATABASE_URL_ASYNC",
    "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
)

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def verify_documents_table():
    async with AsyncSessionLocal() as session:
        # Count total
        result = await session.execute(text("SELECT COUNT(*) FROM documents"))
        total = result.scalar()
        print(f"✅ Total documents in table: {total}")
        
        # Count by category
        result = await session.execute(text("""
            SELECT category, COUNT(*) as count
            FROM documents
            GROUP BY category
            ORDER BY count DESC
        """))
        
        print("\n📁 By category:")
        for row in result.fetchall():
            print(f"   {row[0]}: {row[1]} documents")
        
        # Sample documents
        result = await session.execute(text("""
            SELECT document_id, document_name, category
            FROM documents
            ORDER BY category, document_id
            LIMIT 10
        """))
        
        print("\n📄 Sample:")
        for row in result.fetchall():
            print(f"   {row[0]}: {row[1][:40]}... ({row[2]})")

await verify_documents_table()
await engine.dispose()

✅ Total documents in table: 59

📁 By category:
   Hồ sơ mời thầu: 37 documents
   Mẫu báo cáo: 10 documents
   Câu hỏi thi: 4 documents
   Luật chính: 4 documents
   Thông tư: 2 documents
   Nghị định: 1 documents
   Quyết định: 1 documents

📄 Sample:
   EXAM-Ngân-hàng-câu-hỏi-CC: Ngân hàng câu hỏi CCDT đợt 2... (Câu hỏi thi)
   EXAM-Ngân-hàng-câu-hỏi-th: Ngân hàng câu hỏi thi CCDT đợt 1... (Câu hỏi thi)
   EXAM-NHCH_2692025_dot-2: NHCH 26.9.2025 dot 2- bổ sung... (Câu hỏi thi)
   EXAM-NHCH_30925_bo_sung: NHCH 30.9.25 bo sung theo TB1952 qldt (1... (Câu hỏi thi)
   FORM-01-Phụ-lục: 01. Phụ lục... (Hồ sơ mời thầu)
   FORM-041A-Mẫu-Kế-hoạch-: 04.1A. Mẫu Kế hoạch kiểm tra định kỳ... (Hồ sơ mời thầu)
   FORM-041B-Mẫu-Kế-hoạch-: 04.1B. Mẫu Kế hoạch kiểm tra chi tiết... (Hồ sơ mời thầu)
   FORM-042-Mẫu-Đề-cương-b: 04.2. Mẫu Đề cương báo cáo đấu thầu lựa ... (Hồ sơ mời thầu)
   FORM-043-Mẫu-Báo-cáo-ki: 04.3. Mẫu Báo cáo kiểm tra đối với lựa c... (Hồ sơ mời thầu)
   FORM-044-Mẫu-Kết-luận-k: 04

## Step 3: Build Chunk → Document Mapping

**Thời gian:** 30 phút  
**Input:** metadata files + documents table  
**Output:** chunk_to_document_mapping.json

**Strategy:**
- Sử dụng metadata files (63 files) làm bridge
- Map: chunk_file → source_file → document_id
- Mỗi chunk trong file sẽ có cùng document_id

In [8]:
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text
import os
import json
from pathlib import Path

print("🔨 Building chunk → document mapping...\n")

# Get DATABASE_URL if not already defined
try:
    DATABASE_URL
except NameError:
    DATABASE_URL = os.getenv(
        "DATABASE_URL_ASYNC",
        "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
    )

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

# Load documents from database
async def load_documents():
    async with AsyncSessionLocal() as session:
        result = await session.execute(text("""
            SELECT document_id, source_file, file_name
            FROM documents
            ORDER BY document_id
        """))
        return result.fetchall()

db_documents = await load_documents()
print(f"✅ Loaded {len(db_documents)} documents from database")

# Load metadata files (now in project root)
METADATA_DIR = Path("data/processed/metadata")
metadata_files = list(METADATA_DIR.glob("*.json"))
print(f"✅ Found {len(metadata_files)} metadata files")

# Build mapping: chunk_file → document_id
chunk_mapping = {}  # {chunk_file_path: document_id}
source_to_doc = {doc[1]: doc[0] for doc in db_documents}  # {source_file: document_id}

for meta_file in metadata_files:
    with open(meta_file, 'r', encoding='utf-8') as f:
        metadata = json.load(f)
    
    source_file = metadata.get("source_file")
    chunk_file = metadata.get("output_file")
    
    if source_file in source_to_doc:
        document_id = source_to_doc[source_file]
        chunk_mapping[chunk_file] = document_id

print(f"\n✅ Built mapping for {len(chunk_mapping)} chunk files")
print(f"\n📋 Sample mappings:")
for i, (chunk_file, doc_id) in enumerate(list(chunk_mapping.items())[:5]):
    chunk_name = Path(chunk_file).name
    print(f"   {chunk_name} → {doc_id}")

# Save mapping (project root)
mapping_output = Path("data/processed/chunk_to_document_mapping.json")
with open(mapping_output, 'w', encoding='utf-8') as f:
    json.dump(chunk_mapping, f, ensure_ascii=False, indent=2)

print(f"\n💾 Saved to: {mapping_output}")

await engine.dispose()


🔨 Building chunk → document mapping...



✅ Loaded 59 documents from database
✅ Found 63 metadata files

✅ Built mapping for 55 chunk files

📋 Sample mappings:
   05._Mẫu_Báo_cáo_đấu_thầu.jsonl → FORM-05-Mẫu-Báo-cáo-đấu-
   14A._Mẫu_BCĐG_PTV_HH_XL_hop_hop_TBYT_CGTT._quy_trình_1_1_tui.jsonl → TEMPLATE-BC-14A
   4._Mẫu_số_4B_E-HSMT_hàng_hóa_2_túi.jsonl → FORM-HSYC-HANGHOA-4
   8._Mẫu_số_8B._E-HSMT_EC_02_túi.jsonl → FORM-8-Mẫu-số-8B-E-HSMT
   01D._Mẫu_HSYC_Tư_vấn.jsonl → FORM-HSYC-TUVAN-01D

💾 Saved to: data/processed/chunk_to_document_mapping.json


## Step 4: Update Chunk Document IDs in Database

**Thời gian:** 2 giờ  
**Input:** chunk_to_document_mapping.json  
**Output:** Updated 4708 chunks with new document_ids

**Strategy:**
1. Read chunks từ chunk files
2. Generate new document_id và chunk_id
3. Update database chunks (keep embeddings!)
4. Add source_file metadata

**⚠️ Important:** Không động vào embeddings, chỉ update cmetadata

In [9]:
import asyncio
import asyncpg
import os
import json
from pathlib import Path

print("🚀 Updating chunk document_ids in database...\n")

# Load mapping
with open(mapping_output, 'r', encoding='utf-8') as f:
    chunk_mapping = json.load(f)

print(f"📂 Loaded mapping for {len(chunk_mapping)} chunk files")

# Load metadata for source_file info
metadata_map = {}  # {chunk_file: metadata}
for meta_file in metadata_files:
    with open(meta_file, 'r', encoding='utf-8') as f:
        metadata = json.load(f)
    chunk_file = metadata.get("output_file")
    if chunk_file:
        metadata_map[chunk_file] = metadata

print(f"📋 Loaded metadata for {len(metadata_map)} files\n")

# Database connection params
DB_HOST = os.getenv("DB_HOST", "localhost")
DB_PORT = os.getenv("DB_PORT", "5432")
DB_NAME = os.getenv("DB_NAME", "rag_bidding_v2")
DB_USER = os.getenv("DB_USER", "sakana")
DB_PASSWORD = os.getenv("DB_PASSWORD", "sakana123")

async def update_chunks_v5():
    """Version 5 - Pure asyncpg (no SQLAlchemy) - simpler parameter binding"""
    total_updated = 0
    total_chunks = 0
    errors = []
    
    # Connect directly with asyncpg
    conn = await asyncpg.connect(
        host=DB_HOST,
        port=DB_PORT,
        database=DB_NAME,
        user=DB_USER,
        password=DB_PASSWORD
    )
    
    try:
        print(f"✅ Connected to database\n")
        
        for chunk_file, new_doc_id in chunk_mapping.items():
            # Get metadata
            metadata = metadata_map.get(chunk_file)
            if not metadata:
                errors.append(f"No metadata for {chunk_file}")
                continue
            
            source_file = metadata.get("source_file")
            chunk_file_path = Path(chunk_file)
            
            if not chunk_file_path.exists():
                errors.append(f"Chunk file not found: {chunk_file}")
                continue
            
            # Read chunks from file
            with open(chunk_file_path, 'r', encoding='utf-8') as f:
                for line in f:
                    try:
                        chunk = json.loads(line)
                        old_chunk_id = chunk.get("chunk_id")
                        
                        if not old_chunk_id:
                            continue
                        
                        # Generate new chunk_id
                        parts = old_chunk_id.split("_")
                        if len(parts) >= 3:
                            chunk_suffix = "_".join(parts[2:])
                        else:
                            chunk_suffix = old_chunk_id
                        
                        new_chunk_id = f"{new_doc_id}_{chunk_suffix}"
                        
                        # Update with asyncpg - uses $1, $2, $3, $4 positional params
                        query = """
                            UPDATE langchain_pg_embedding
                            SET cmetadata = jsonb_set(
                                jsonb_set(
                                    jsonb_set(
                                        cmetadata,
                                        '{document_id}',
                                        to_jsonb($1::text)
                                    ),
                                    '{chunk_id}',
                                    to_jsonb($2::text)
                                ),
                                '{source_file}',
                                to_jsonb($3::text)
                            )
                            WHERE cmetadata->>'chunk_id' = $4
                        """
                        
                        result = await conn.execute(
                            query,
                            new_doc_id,
                            new_chunk_id,
                            source_file,
                            old_chunk_id
                        )
                        
                        # result format: "UPDATE 1" or "UPDATE 0"
                        rows_affected = int(result.split()[-1])
                        if rows_affected > 0:
                            total_updated += 1
                        
                        total_chunks += 1
                        
                        # Progress update
                        if total_chunks % 100 == 0:
                            print(f"  📊 Processed {total_chunks} chunks, updated {total_updated}...")
                    
                    except Exception as e:
                        errors.append(f"Error processing chunk in {chunk_file}: {e}")
        
        print(f"\n✅ All chunks processed!")
        
    finally:
        await conn.close()
        print(f"✅ Database connection closed\n")
    
    return total_updated, total_chunks, errors

# Run the update
updated, processed, errors = await update_chunks_v5()

print(f"\n{'='*60}")
print(f"✅ Processed {processed} chunks")
print(f"✅ Updated {updated} chunks in database")

if errors:
    print(f"\n⚠️  Errors: {len(errors)}")
    for err in errors[:5]:
        print(f"   {err}")
    if len(errors) > 5:
        print(f"   ... and {len(errors) - 5} more errors")


🚀 Updating chunk document_ids in database...

📂 Loaded mapping for 55 chunk files
📋 Loaded metadata for 63 files

✅ Connected to database

  📊 Processed 100 chunks, updated 0...
  📊 Processed 100 chunks, updated 0...
  📊 Processed 200 chunks, updated 0...
  📊 Processed 200 chunks, updated 0...
  📊 Processed 300 chunks, updated 0...
  📊 Processed 300 chunks, updated 0...
  📊 Processed 400 chunks, updated 0...
  📊 Processed 400 chunks, updated 0...
  📊 Processed 500 chunks, updated 0...
  📊 Processed 500 chunks, updated 0...
  📊 Processed 600 chunks, updated 0...
  📊 Processed 600 chunks, updated 0...
  📊 Processed 700 chunks, updated 0...
  📊 Processed 700 chunks, updated 0...
  📊 Processed 800 chunks, updated 0...
  📊 Processed 800 chunks, updated 0...
  📊 Processed 900 chunks, updated 0...
  📊 Processed 900 chunks, updated 0...
  📊 Processed 1000 chunks, updated 0...
  📊 Processed 1000 chunks, updated 0...
  📊 Processed 1100 chunks, updated 0...
  📊 Processed 1100 chunks, updated 0...

---

# NGÀY 2: Verification + API Updates

**Thời gian:** 4 giờ  
**Mục tiêu:** Verify migration + Update API endpoints

## Step 5: Verify Migration

**Thời gian:** 1 giờ  
**Kiểm tra:**
1. Số lượng documents trong database
2. source_file được populate
3. document_id format đúng
4. chunk_id format đúng

In [10]:
print("🔍 Verifying migration...\n")

# Get DATABASE_URL if not already defined
try:
    DATABASE_URL
except NameError:
    DATABASE_URL = os.getenv(
        "DATABASE_URL_ASYNC",
        "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
    )

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def verify_migration():
    async with AsyncSessionLocal() as session:
        # 1. Count distinct document_ids
        result = await session.execute(text("""
            SELECT COUNT(DISTINCT cmetadata->>'document_id')
            FROM langchain_pg_embedding
        """))
        doc_count = result.scalar()
        print(f"✅ Distinct document_ids: {doc_count}")
        print(f"   Expected: 70, Actual: {doc_count}")
        
        # 2. Check source_file populated
        result = await session.execute(text("""
            SELECT 
                COUNT(*) FILTER (WHERE cmetadata->>'source_file' IS NOT NULL) as with_source,
                COUNT(*) as total
            FROM langchain_pg_embedding
        """))
        row = result.fetchone()
        print(f"\n✅ Source file tracking:")
        print(f"   With source_file: {row[0]} / {row[1]}")
        print(f"   Coverage: {row[0]/row[1]*100:.1f}%")
        
        # 3. Sample document_ids
        result = await session.execute(text("""
            SELECT 
                cmetadata->>'document_id' as doc_id,
                cmetadata->>'source_file' as source,
                COUNT(*) as chunks
            FROM langchain_pg_embedding
            GROUP BY cmetadata->>'document_id', cmetadata->>'source_file'
            ORDER BY doc_id
            LIMIT 15
        """))
        
        print(f"\n📋 Sample documents in database:")
        for row in result.fetchall():
            doc_id = row[0]
            source = row[1]
            chunks = row[2]
            source_display = source.split('/')[-1][:40] if source else "(NULL)"
            print(f"   {doc_id:30} | {chunks:4} chunks | {source_display}")
        
        # 4. Sample chunk_ids
        result = await session.execute(text("""
            SELECT cmetadata->>'chunk_id'
            FROM langchain_pg_embedding
            LIMIT 5
        """))
        
        print(f"\n📄 Sample chunk_ids:")
        for row in result.fetchall():
            print(f"   {row[0]}")
        
        # 5. Check for old format (should be 0)
        result = await session.execute(text("""
            SELECT COUNT(*)
            FROM langchain_pg_embedding
            WHERE cmetadata->>'document_id' LIKE '%untitled%'
        """))
        old_format = result.scalar()
        print(f"\n⚠️  Old format document_ids: {old_format}")
        if old_format > 0:
            print(f"   Warning: Still have {old_format} chunks with old format!")

await verify_migration()
await engine.dispose()

🔍 Verifying migration...

✅ Distinct document_ids: 25
   Expected: 70, Actual: 25

✅ Source file tracking:
   With source_file: 4705 / 4708
   Coverage: 99.9%

📋 Sample documents in database:
   FORM-05-Mẫu-Báo-cáo-đấu-       |  130 chunks | 05. Mẫu Báo cáo đấu thầu.docx
   FORM-10-Mẫu-số-10A_E-HSM       |    6 chunks | 10. Mẫu số 10A_E-HSMT_EPC 01 túi.docx
   FORM-10-Mẫu-số-10B_E-HSM       |  104 chunks | 10. Mẫu số 10B_E-HSMT_EPC 2 túi.docx
   FORM-11-Mẫu-số-11A_E-HSM       |   23 chunks | 11. Mẫu số 11A_E-HSMT_Máy đặt máy mượn 1
   FORM-11-Mẫu-số-11B_E-HSM       |    9 chunks | 11. Mẫu số 11B_E-HSMT_Máy đặt máy mượn 2
   FORM-15-Phu-luc                |  153 chunks | 15. Phu luc.docx
   FORM-7-Mẫu-số-7A-E-HSMT        |    9 chunks | 7. Mẫu số 7A. E-HSMT_EP qua mạng 01 túi.
   FORM-7-Mẫu-số-7B-E-HSMT        |  264 chunks | 7. Mẫu số 7B. E-HSMT_EP qua mạng 2 túi.d
   FORM-8-Mẫu-số-8B-E-HSMT        |  166 chunks | 8. Mẫu số 8B. E-HSMT_EC 02 túi.docx
   FORM-9Mẫu-số-9A_E-HSMT_P       | 

In [11]:
import asyncio
import asyncpg
import os
from collections import defaultdict

print("🔍 DETAILED MIGRATION ANALYSIS\n")
print("="*60)

# Database connection params
DB_HOST = os.getenv("DB_HOST", "localhost")
DB_PORT = os.getenv("DB_PORT", "5432")
DB_NAME = os.getenv("DB_NAME", "rag_bidding_v2")
DB_USER = os.getenv("DB_USER", "sakana")
DB_PASSWORD = os.getenv("DB_PASSWORD", "sakana123")

async def analyze_migration():
    conn = await asyncpg.connect(
        host=DB_HOST,
        port=DB_PORT,
        database=DB_NAME,
        user=DB_USER,
        password=DB_PASSWORD
    )
    
    try:
        # 1. Get all distinct document_ids from vector DB
        result = await conn.fetch("""
            SELECT 
                cmetadata->>'document_id' as doc_id,
                COUNT(*) as chunk_count
            FROM langchain_pg_embedding
            GROUP BY cmetadata->>'document_id'
            ORDER BY doc_id
        """)
        
        print(f"📊 Vector Database Analysis:")
        print(f"   Total distinct document_ids: {len(result)}\n")
        
        # Group by prefix to see patterns
        by_prefix = defaultdict(list)
        for row in result:
            doc_id = row['doc_id']
            if doc_id:
                prefix = doc_id.split('-')[0]
                by_prefix[prefix].append((doc_id, row['chunk_count']))
        
        print(f"📋 Document IDs by prefix:")
        for prefix, docs in sorted(by_prefix.items()):
            print(f"\n   {prefix}: {len(docs)} documents")
            for doc_id, count in docs[:5]:  # Show first 5
                print(f"      {doc_id:40} ({count:4} chunks)")
            if len(docs) > 5:
                print(f"      ... and {len(docs) - 5} more")
        
        # 2. Check mapping file
        print(f"\n{'='*60}")
        print(f"📂 Mapping File Analysis:\n")
        
        import json
        from pathlib import Path
        
        mapping_file = Path("data/processed/chunk_to_document_mapping.json")
        with open(mapping_file, 'r') as f:
            chunk_mapping = json.load(f)
        
        print(f"   Total chunk files in mapping: {len(chunk_mapping)}")
        print(f"   Unique document_ids in mapping: {len(set(chunk_mapping.values()))}")
        
        # Count by document_id
        doc_counts = defaultdict(int)
        for doc_id in chunk_mapping.values():
            doc_counts[doc_id] += 1
        
        print(f"\n   Top 10 document_ids by chunk file count:")
        for doc_id, count in sorted(doc_counts.items(), key=lambda x: x[1], reverse=True)[:10]:
            print(f"      {doc_id:40} ({count} chunk files)")
        
        # 3. Find missing documents
        print(f"\n{'='*60}")
        print(f"🔍 Missing Documents Analysis:\n")
        
        mapping_doc_ids = set(chunk_mapping.values())
        vector_doc_ids = {row['doc_id'] for row in result if row['doc_id']}
        
        missing_in_vector = mapping_doc_ids - vector_doc_ids
        extra_in_vector = vector_doc_ids - mapping_doc_ids
        
        if missing_in_vector:
            print(f"   ⚠️  {len(missing_in_vector)} document_ids in mapping but NOT in vector DB:")
            for doc_id in sorted(missing_in_vector)[:10]:
                # Find which chunk files map to this
                chunk_files = [cf for cf, did in chunk_mapping.items() if did == doc_id]
                print(f"      {doc_id}")
                print(f"         Chunk files: {len(chunk_files)}")
                if chunk_files:
                    print(f"         Example: {Path(chunk_files[0]).name}")
        else:
            print(f"   ✅ All mapping document_ids found in vector DB")
        
        if extra_in_vector:
            print(f"\n   ℹ️  {len(extra_in_vector)} document_ids in vector DB but NOT in mapping:")
            for doc_id in sorted(extra_in_vector)[:5]:
                print(f"      {doc_id}")
        
        # 4. Check which chunk files were actually processed
        print(f"\n{'='*60}")
        print(f"📝 Chunk File Processing Status:\n")
        
        # Get sample of old chunk_ids still in DB
        old_chunks = await conn.fetch("""
            SELECT DISTINCT 
                cmetadata->>'chunk_id' as chunk_id,
                cmetadata->>'document_id' as doc_id
            FROM langchain_pg_embedding
            WHERE cmetadata->>'chunk_id' LIKE '%untitled%'
               OR cmetadata->>'chunk_id' LIKE '%bidding_%'
               OR cmetadata->>'chunk_id' LIKE '%law_%'
            LIMIT 20
        """)
        
        if old_chunks:
            print(f"   ⚠️  Found {len(old_chunks)} chunks with OLD format chunk_ids:")
            for row in old_chunks[:10]:
                print(f"      {row['chunk_id']:50} → {row['doc_id']}")
        else:
            print(f"   ✅ No old format chunk_ids found")
        
    finally:
        await conn.close()

await analyze_migration()

print(f"\n{'='*60}")
print(f"✅ Analysis complete!")
print(f"{'='*60}")

🔍 DETAILED MIGRATION ANALYSIS

📊 Vector Database Analysis:
   Total distinct document_ids: 25

📋 Document IDs by prefix:

   FORM: 16 documents
      FORM-05-Mẫu-Báo-cáo-đấu-                 ( 130 chunks)
      FORM-10-Mẫu-số-10A_E-HSM                 (   6 chunks)
      FORM-10-Mẫu-số-10B_E-HSM                 ( 104 chunks)
      FORM-11-Mẫu-số-11A_E-HSM                 (  23 chunks)
      FORM-11-Mẫu-số-11B_E-HSM                 (   9 chunks)
      ... and 11 more

   LUA: 4 documents
      LUA-57-2024-QH15                         ( 271 chunks)
      LUA-90-2025-QH15                         (  78 chunks)
      LUA-HỢP-NHẤT-126-2025-về                 ( 617 chunks)
      LUA-Luat-dau-thau-2023                   ( 188 chunks)

   ND: 1 documents
      ND-214-4.8.-CP                           ( 595 chunks)

   QD: 1 documents
      QD-1667-BYT                              (   5 chunks)

   TEMPLATE: 1 documents
      TEMPLATE-BC-14A                          ( 510 chunks)

   TT: 2 docum

## 🎯 Phân Tích Nguyên Nhân & Giải Pháp

### ❌ **Vấn Đề Phát Hiện**: 31 Documents Chưa Được Import Vào Database

**Kết Quả Phân Tích:**
1. ✅ **Chunk files**: Có 63 files trong `data/processed/chunks/`
2. ✅ **Mapping file**: Có 55 chunk files được map sang document_ids
3. ✅ **Format trong database**: Chunks ĐÃ có format mới (`FORM-05-..._section_0000`)
4. ❌ **Thiếu chunks**: 31 document_ids có trong mapping nhưng KHÔNG có trong vector DB

**Nguyên Nhân:**
- 31 documents bị thiếu **chưa bao giờ được import** vào vector database
- Đây KHÔNG phải lỗi migration (chunks đã có format đúng rồi)
- Đây là **lỗ hổng data ingestion** từ giai đoạn xử lý ban đầu

**Bằng Chứng:**
```sql
-- Database đã có format mới:
LUA-57-2024-QH15_dieu_0143
FORM-05-Mẫu-Báo-cáo-đấu-_section_0000
FORM-HSYC-TUVAN-01D_form_0004

-- Nhưng documents bị thiếu như FORM-01-Phụ-lục có 0 chunks trong DB
```

### ✅ **Các Phương Án Giải Quyết:**

**Phương Án 1: Hoàn Thành Migration (Khuyên Dùng)** 🌟
- **Là gì**: Import lại 31 chunk files bị thiếu vào vector database
- **Cách thực hiện**: Dùng `scripts/import_processed_chunks.py` để import các files còn thiếu
- **Kết quả**: Tăng từ 25 → 55 documents (trạng thái mong muốn)
- **Thời gian**: 10-15 phút

**Phương Án 2: Chỉ Cập Nhật Mapping**
- **Là gì**: Xóa 31 documents bị thiếu khỏi mapping file
- **Cách thực hiện**: Cập nhật `chunk_to_document_mapping.json` để chỉ chứa 25 documents
- **Kết quả**: Giữ nguyên trạng thái hiện tại (25 documents), cập nhật mapping cho khớp
- **Thời gian**: 2 phút

**Khuyến Nghị**: **Phương Án 1** - Hoàn thành việc import data để có đủ 55 documents.

---

### 📋 Các Bước Tiếp Theo:

1. **Xác định các chunk files bị thiếu** (31 files)
2. **Chạy import script** cho các files thiếu
3. **Verify**: Phải có 55 documents trong database
4. **Tiếp tục** với Steps 6-7

In [12]:
import asyncio
import asyncpg
import os
import json
from pathlib import Path

print("🔍 Xác Định Các Chunk Files Bị Thiếu Để Import\n")
print("="*60)

# Kết nối database
DB_HOST = os.getenv("DB_HOST", "localhost")
DB_PORT = os.getenv("DB_PORT", "5432")
DB_NAME = os.getenv("DB_NAME", "rag_bidding_v2")
DB_USER = os.getenv("DB_USER", "sakana")
DB_PASSWORD = os.getenv("DB_PASSWORD", "sakana123")

async def identify_missing_files():
    conn = await asyncpg.connect(
        host=DB_HOST, port=DB_PORT, database=DB_NAME,
        user=DB_USER, password=DB_PASSWORD
    )
    
    try:
        # Lấy tất cả document_ids hiện có trong vector DB
        result = await conn.fetch("""
            SELECT DISTINCT cmetadata->>'document_id' as doc_id
            FROM langchain_pg_embedding
        """)
        vector_doc_ids = {row['doc_id'] for row in result if row['doc_id']}
        
        # Load mapping file
        mapping_file = Path("data/processed/chunk_to_document_mapping.json")
        with open(mapping_file, 'r') as f:
            chunk_mapping = json.load(f)
        
        mapping_doc_ids = set(chunk_mapping.values())
        
        # Tìm documents bị thiếu
        missing_doc_ids = mapping_doc_ids - vector_doc_ids
        
        print(f"📊 Tổng Kết:")
        print(f"   Documents trong mapping: {len(mapping_doc_ids)}")
        print(f"   Documents trong vector DB: {len(vector_doc_ids)}")
        print(f"   Documents bị thiếu: {len(missing_doc_ids)}\n")
        
        # Tìm chunk files cho documents bị thiếu
        missing_files = []
        for chunk_file, doc_id in chunk_mapping.items():
            if doc_id in missing_doc_ids:
                chunk_path = Path(chunk_file)
                if chunk_path.exists():
                    # Đếm chunks trong file
                    chunk_count = sum(1 for line in open(chunk_path, 'r'))
                    missing_files.append({
                        'document_id': doc_id,
                        'chunk_file': str(chunk_path),
                        'file_name': chunk_path.name,
                        'chunk_count': chunk_count,
                        'exists': True
                    })
                else:
                    missing_files.append({
                        'document_id': doc_id,
                        'chunk_file': str(chunk_path),
                        'file_name': chunk_path.name,
                        'chunk_count': 0,
                        'exists': False
                    })
        
        # Sắp xếp theo document_id
        missing_files.sort(key=lambda x: x['document_id'])
        
        print(f"📂 Các Chunk Files Bị Thiếu ({len(missing_files)}):\n")
        
        total_chunks = 0
        existing_files = 0
        
        for i, file_info in enumerate(missing_files, 1):
            status = "✅" if file_info['exists'] else "❌"
            print(f"   [{i:2d}] {status} {file_info['document_id']}")
            print(f"        File: {file_info['file_name']}")
            print(f"        Số chunks: {file_info['chunk_count']}")
            print()
            
            if file_info['exists']:
                total_chunks += file_info['chunk_count']
                existing_files += 1
        
        print(f"{'='*60}")
        print(f"✅ Files tồn tại: {existing_files}/{len(missing_files)}")
        print(f"📊 Tổng chunks cần import: {total_chunks:,}")
        print(f"{'='*60}\n")
        
        # Lưu danh sách để import
        output_file = Path("data/processed/missing_chunk_files.json")
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(missing_files, f, ensure_ascii=False, indent=2)
        
        print(f"💾 Đã lưu danh sách files bị thiếu vào: {output_file}")
        
        return missing_files
    
    finally:
        await conn.close()

missing_files = await identify_missing_files()

🔍 Xác Định Các Chunk Files Bị Thiếu Để Import

📊 Tổng Kết:
   Documents trong mapping: 55
   Documents trong vector DB: 25
   Documents bị thiếu: 31

📂 Các Chunk Files Bị Thiếu (31):

   [ 1] ✅ FORM-01-Phụ-lục
        File: 01._Phụ_lục.jsonl
        Số chunks: 48

   [ 2] ✅ FORM-041A-Mẫu-Kế-hoạch-
        File: 04.1A._Mẫu_Kế_hoạch_kiểm_tra_định_kỳ.jsonl
        Số chunks: 1

   [ 3] ✅ FORM-041B-Mẫu-Kế-hoạch-
        File: 04.1B._Mẫu_Kế_hoạch_kiểm_tra_chi_tiết.jsonl
        Số chunks: 10

   [ 4] ✅ FORM-042-Mẫu-Đề-cương-b
        File: 04.2._Mẫu_Đề_cương_báo_cáo_đấu_thầu_lựa_chọn_nhà_thầu,_NĐT.jsonl
        Số chunks: 10

   [ 5] ✅ FORM-043-Mẫu-Báo-cáo-ki
        File: 04.3._Mẫu_Báo_cáo_kiểm_tra_đối_với_lựa_chọn_nhà_thầu,_NĐT.jsonl
        Số chunks: 8

   [ 6] ✅ FORM-044-Mẫu-Kết-luận-k
        File: 04.4._Mẫu_Kết_luận_kiểm_tra_đối_với_lựa_chọn_nhà_thầu,_NĐT.jsonl
        Số chunks: 4

   [ 7] ✅ FORM-045-Mẫu-Báo-cáo-ph
        File: 04.5._Mẫu_Báo_cáo_phản_hồi_về_Kết_luận_kiểm_tra.jsonl


### 🚀 Import Các Chunk Files Bị Thiếu

**Chiến Lược:**
1. Đọc các chunk files bị thiếu (đã có format đúng rồi)
2. Import trực tiếp vào vector database cùng với embeddings
3. Verify: Phải tăng từ 25 → 55 documents

**⚠️ Quan Trọng:**
- Chunks trong files ĐÃ có format đúng (`document_id` và `chunk_id`)
- Chúng ta chỉ cần import chúng vào vector database
- Sử dụng script import hiện có: `scripts/import_processed_chunks.py`

### ⚠️ Cleanup: Xóa Chunks Import Sai

**Vấn đề phát hiện:**
- Cell 28 lần trước import 767 chunks với `document_id: bidding_untitled` (old format)
- Cần xóa chúng trước khi re-import với document_ids đúng

**Giải pháp:**
- Xóa các chunks có `document_id: bidding_untitled` (hoặc giống old format)
- Re-import với document_ids mới từ mapping

In [ ]:
import asyncio
import asyncpg
import os

print("🗑️ Cleanup: Xóa Chunks Import Sai\n")
print("="*60)

# Kết nối database
DB_HOST = os.getenv("DB_HOST", "localhost")
DB_PORT = os.getenv("DB_PORT", "5432")
DB_NAME = os.getenv("DB_NAME", "rag_bidding_v2")
DB_USER = os.getenv("DB_USER", "sakana")
DB_PASSWORD = os.getenv("DB_PASSWORD", "sakana123")

async def cleanup_wrong_imports():
    conn = await asyncpg.connect(
        host=DB_HOST, port=DB_PORT, database=DB_NAME,
        user=DB_USER, password=DB_PASSWORD
    )
    
    try:
        # Kiểm tra chunks với bidding_untitled (import sai lần trước)
        result = await conn.fetchrow("""
            SELECT COUNT(*) as count
            FROM langchain_pg_embedding
            WHERE cmetadata->>'document_id' = 'bidding_untitled'
        """)
        
        print(f"📊 Chunks với document_id='bidding_untitled': {result['count']}")
        
        if result['count'] > 0:
            print(f"   🗑️ Đang xóa {result['count']} chunks...\n")
            
            # Xóa chunks
            deleted = await conn.execute("""
                DELETE FROM langchain_pg_embedding
                WHERE cmetadata->>'document_id' = 'bidding_untitled'
            """)
            
            print(f"   ✅ Đã xóa: {deleted}")
        else:
            print(f"   ✅ Không có chunks cần xóa")
        
        # Kiểm tra lại sau khi xóa
        result = await conn.fetchrow("""
            SELECT COUNT(DISTINCT cmetadata->>'document_id') as doc_count,
                   COUNT(*) as total_chunks
            FROM langchain_pg_embedding
        """)
        
        print(f"\n📊 Sau khi cleanup:")
        print(f"   Số document_ids: {result['doc_count']}")
        print(f"   Tổng chunks: {result['total_chunks']:,}")
        print(f"\n{'='*60}")
        
    finally:
        await conn.close()

await cleanup_wrong_imports()

In [16]:
import sys
import json
from pathlib import Path
from tqdm import tqdm

# Import các utilities để import
sys.path.insert(0, '/home/sakana/Code/RAG-bidding')

from langchain_core.documents import Document
from langchain_openai import OpenAIEmbeddings
from langchain_postgres import PGVector
from src.config.models import settings

print("🚀 Đang Import Các Chunk Files Bị Thiếu (Với Document ID Mới)\n")
print("="*60)

# Load danh sách files bị thiếu
missing_list_file = Path("data/processed/missing_chunk_files.json")
with open(missing_list_file, 'r') as f:
    missing_files_info = json.load(f)

# Lọc chỉ lấy files tồn tại
files_to_ingest = [f for f in missing_files_info if f['exists']]

print(f"📂 Files cần import: {len(files_to_ingest)}")
print(f"📊 Tổng chunks: {sum(f['chunk_count'] for f in files_to_ingest):,}\n")

# Load metadata để lấy source_file
METADATA_DIR = Path("data/processed/metadata")
metadata_files = list(METADATA_DIR.glob("*.json"))
metadata_by_chunk_file = {}

for meta_file in metadata_files:
    with open(meta_file, 'r', encoding='utf-8') as f:
        metadata = json.load(f)
    chunk_file = metadata.get("output_file")
    if chunk_file:
        metadata_by_chunk_file[chunk_file] = metadata

print(f"📋 Loaded metadata for {len(metadata_by_chunk_file)} files\n")

# Khởi tạo vector store
embeddings = OpenAIEmbeddings(model=settings.embed_model)
vector_store = PGVector(
    embeddings=embeddings,
    collection_name=settings.collection,
    connection=settings.database_url,
    use_jsonb=True,
)

print(f"✅ Đã kết nối vector store: {settings.collection}\n")

# Import từng file
total_ingested = 0
total_errors = 0

for i, file_info in enumerate(files_to_ingest, 1):
    chunk_file = Path(file_info['chunk_file'])
    new_doc_id = file_info['document_id']  # Document ID MỚI từ mapping
    chunk_count = file_info['chunk_count']
    
    # Lấy source_file từ metadata
    metadata_info = metadata_by_chunk_file.get(str(chunk_file))
    source_file = metadata_info.get("source_file") if metadata_info else ""
    
    print(f"[{i}/{len(files_to_ingest)}] Đang import: {new_doc_id}")
    print(f"   File: {chunk_file.name}")
    print(f"   Chunks: {chunk_count}")
    
    try:
        # Load chunks từ file
        chunks = []
        with open(chunk_file, 'r', encoding='utf-8') as f:
            for line in f:
                if line.strip():
                    try:
                        chunk = json.loads(line)
                        chunks.append(chunk)
                    except json.JSONDecodeError:
                        total_errors += 1
        
        # Chuyển sang LangChain Documents VÀ CẬP NHẬT document_id mới
        documents = []
        for chunk in chunks:
            content = chunk.get("content", "")
            old_chunk_id = chunk.get("chunk_id", "")
            
            # Tạo chunk_id mới từ new_doc_id
            # Format cũ: bidding_untitled_form_0000
            # Format mới: FORM-01-Phụ-lục_form_0000
            if old_chunk_id:
                parts = old_chunk_id.split("_")
                if len(parts) >= 3:
                    # Lấy suffix (form_0000, section_0000, etc.)
                    chunk_suffix = "_".join(parts[2:])
                else:
                    chunk_suffix = old_chunk_id
                new_chunk_id = f"{new_doc_id}_{chunk_suffix}"
            else:
                new_chunk_id = f"{new_doc_id}_chunk_{len(documents)}"
            
            metadata = {
                "chunk_id": new_chunk_id,  # ✅ Chunk ID MỚI
                "document_id": new_doc_id,  # ✅ Document ID MỚI
                "document_type": chunk.get("document_type"),
                "hierarchy": json.dumps(chunk.get("hierarchy", []), ensure_ascii=False),
                "level": chunk.get("level"),
                "section_title": chunk.get("section_title", ""),
                "char_count": chunk.get("char_count", len(content)),
                "chunk_index": chunk.get("chunk_index", 0),
                "total_chunks": chunk.get("total_chunks", 1),
                "is_complete_unit": chunk.get("is_complete_unit", True),
                "has_table": chunk.get("has_table", False),
                "has_list": chunk.get("has_list", False),
                "source_file": source_file,  # ✅ Source file từ metadata
            }
            documents.append(Document(page_content=content, metadata=metadata))
        
        # Thêm vào vector store (theo batch)
        if documents:
            batch_size = 50
            for j in range(0, len(documents), batch_size):
                batch = documents[j:j+batch_size]
                vector_store.add_documents(batch)
            
            total_ingested += len(documents)
            print(f"   ✅ Đã import {len(documents)} chunks với document_id: {new_doc_id}\n")
        else:
            print(f"   ⚠️  Không tìm thấy chunks hợp lệ\n")
    
    except Exception as e:
        print(f"   ❌ Lỗi: {e}\n")
        total_errors += 1

print(f"{'='*60}")
print(f"✅ Hoàn Thành Import!")
print(f"   Files đã xử lý: {len(files_to_ingest)}")
print(f"   Chunks đã import: {total_ingested:,}")
if total_errors > 0:
    print(f"   ⚠️  Lỗi: {total_errors}")
print(f"{'='*60}")

🚀 Đang Import Các Chunk Files Bị Thiếu (Với Document ID Mới)

📂 Files cần import: 31
📊 Tổng chunks: 767

📋 Loaded metadata for 63 files

✅ Đã kết nối vector store: docs

[1/31] Đang import: FORM-01-Phụ-lục
   File: 01._Phụ_lục.jsonl
   Chunks: 48
   ✅ Đã import 48 chunks với document_id: FORM-01-Phụ-lục

[2/31] Đang import: FORM-041A-Mẫu-Kế-hoạch-
   File: 04.1A._Mẫu_Kế_hoạch_kiểm_tra_định_kỳ.jsonl
   Chunks: 1
   ✅ Đã import 48 chunks với document_id: FORM-01-Phụ-lục

[2/31] Đang import: FORM-041A-Mẫu-Kế-hoạch-
   File: 04.1A._Mẫu_Kế_hoạch_kiểm_tra_định_kỳ.jsonl
   Chunks: 1
   ✅ Đã import 1 chunks với document_id: FORM-041A-Mẫu-Kế-hoạch-

[3/31] Đang import: FORM-041B-Mẫu-Kế-hoạch-
   File: 04.1B._Mẫu_Kế_hoạch_kiểm_tra_chi_tiết.jsonl
   Chunks: 10
   ✅ Đã import 1 chunks với document_id: FORM-041A-Mẫu-Kế-hoạch-

[3/31] Đang import: FORM-041B-Mẫu-Kế-hoạch-
   File: 04.1B._Mẫu_Kế_hoạch_kiểm_tra_chi_tiết.jsonl
   Chunks: 10
   ✅ Đã import 10 chunks với document_id: FORM-041B-Mẫu-Kế-hoạc

In [17]:
import asyncio
import asyncpg
import os

print("🔍 Kiểm Tra Kết Quả Import\n")
print("="*60)

# Kết nối database
DB_HOST = os.getenv("DB_HOST", "localhost")
DB_PORT = os.getenv("DB_PORT", "5432")
DB_NAME = os.getenv("DB_NAME", "rag_bidding_v2")
DB_USER = os.getenv("DB_USER", "sakana")
DB_PASSWORD = os.getenv("DB_PASSWORD", "sakana123")

async def verify_ingestion():
    conn = await asyncpg.connect(
        host=DB_HOST, port=DB_PORT, database=DB_NAME,
        user=DB_USER, password=DB_PASSWORD
    )
    
    try:
        # Đếm distinct document_ids
        stats = await conn.fetchrow("""
            SELECT COUNT(DISTINCT cmetadata->>'document_id') as doc_count,
                   COUNT(*) as total_chunks
            FROM langchain_pg_embedding
        """)
        
        print(f"📊 Sau Khi Import:")
        print(f"   Số document_ids: {stats['doc_count']}")
        print(f"   Tổng chunks: {stats['total_chunks']:,}")
        print()
        
        # Kiểm tra các documents trước đây bị thiếu
        missing_doc_ids = [
            'FORM-01-Phụ-lục', 'FORM-041A-Mẫu-Kế-hoạch-', 'FORM-041B-Mẫu-Kế-hoạch-',
            'FORM-042-Mẫu-Đề-cương-b', 'FORM-043-Mẫu-Báo-cáo-ki'
        ]
        
        print(f"🔍 Kiểm tra các documents trước đây bị thiếu:\n")
        for doc_id in missing_doc_ids[:5]:
            result = await conn.fetchrow("""
                SELECT COUNT(*) as count
                FROM langchain_pg_embedding
                WHERE cmetadata->>'document_id' = $1
            """, doc_id)
            
            status = "✅" if result['count'] > 0 else "❌"
            print(f"   {status} {doc_id:40} : {result['count']:4} chunks")
        
        print(f"\n{'='*60}")
        
        # So sánh Mong đợi vs Thực tế
        expected_docs = 55
        actual_docs = stats['doc_count']
        
        if actual_docs >= expected_docs:
            print(f"✅ THÀNH CÔNG: Đã có đầy đủ {actual_docs}/{expected_docs} documents!")
        else:
            print(f"⚠️  CHƯA ĐỦ: {actual_docs}/{expected_docs} documents")
            print(f"   Còn thiếu: {expected_docs - actual_docs} documents")
        
        print(f"{'='*60}")
        
    finally:
        await conn.close()

await verify_ingestion()

🔍 Kiểm Tra Kết Quả Import

📊 Sau Khi Import:
   Số document_ids: 57
   Tổng chunks: 6,242

🔍 Kiểm tra các documents trước đây bị thiếu:

   ✅ FORM-01-Phụ-lục                          :   48 chunks
   ✅ FORM-041A-Mẫu-Kế-hoạch-                  :    1 chunks
   ✅ FORM-041B-Mẫu-Kế-hoạch-                  :   10 chunks
   ✅ FORM-042-Mẫu-Đề-cương-b                  :   10 chunks
   ✅ FORM-043-Mẫu-Báo-cáo-ki                  :    8 chunks

✅ THÀNH CÔNG: Đã có đầy đủ 57/55 documents!


---

## 📋 Tổng Kết Tình Trạng Migration

### ✅ Những Gì Chúng Ta Phát Hiện:

**Vấn Đề Ban Đầu:**
- Mong đợi: 55 documents được migrate
- Thực tế: Chỉ tìm thấy 25 documents
- Khoảng cách: 31 documents bị thiếu

**Phân Tích Nguyên Nhân:**
1. ✅ **Script Migration Step 4**: Thực ra đã HOẠT ĐỘNG ĐÚNG!
   - Cập nhật được các chunks có trong database
   - Chunks đã có format `document_id` và `chunk_id` đúng rồi
   
2. ❌ **Vấn Đề Thực Sự**: **Lỗ Hổng Import Data**
   - 31 chunk files tồn tại trên disk nhưng **chưa bao giờ được import** vào vector database
   - Điều này xảy ra trong giai đoạn xử lý data ban đầu (trước khi migration)
   - Migration không thể cập nhật chunks không tồn tại trong database

**Bằng Chứng:**
```
Chunks trong database đã có FORMAT MỚI:
- LUA-57-2024-QH15_dieu_0143
- FORM-05-Mẫu-Báo-cáo-đấu-_section_0000
- FORM-HSYC-TUVAN-01D_form_0004

Documents bị thiếu (31):
- FORM-01-Phụ-lục → chunk file tồn tại, 0 chunks trong DB
- FORM-041A-Mẫu-Kế-hoạch- → chunk file tồn tại, 0 chunks trong DB
- ... và còn 29 documents nữa
```

---

### 🎯 Kế Hoạch Hành Động:

**Giai Đoạn 1: Hoàn Thành Import Data** (Hiện tại)
1. ✅ Xác định 31 chunk files bị thiếu
2. 🔄 Import các files thiếu vào vector database (chạy cell phía trên)
3. ✅ Verify: Phải đạt 55 documents

**Giai Đoạn 2: Hoàn Thành Các Bước Migration**
1. Chạy Step 6: Cập nhật `documents.total_chunks`
2. Chạy Step 7: Tạo báo cáo tổng kết migration

---

### 🚀 Hành Động Tiếp Theo:

**Chạy các cells phía trên theo thứ tự:**
1. **Xác Định Files Bị Thiếu** → Hiển thị 31 files cần import
2. **Import Files Bị Thiếu** → Thêm vào vector database (~10-15 phút, cần OpenAI API)
3. **Kiểm Tra Kết Quả Import** → Xác nhận có đủ 55 documents ✅
4. **Tiếp Tục Step 6** → Cập nhật số lượng chunks trong bảng documents
5. **Step 7** → Tạo báo cáo migration cuối cùng

Script migration hoạt động đúng - chúng ta chỉ cần hoàn thành việc import data trước!

## Step 6: Update documents.total_chunks

**Thời gian:** 30 phút  
**Mục đích:** Sync chunk counts từ database vào documents table

In [18]:
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text
import os

print("🔄 Cập Nhật documents.total_chunks...\n")

# Get DATABASE_URL if not already defined
try:
    DATABASE_URL
except NameError:
    DATABASE_URL = os.getenv(
        "DATABASE_URL_ASYNC",
        "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
    )

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def update_chunk_counts():
    async with AsyncSessionLocal() as session:
        # Lấy tất cả documents
        result = await session.execute(text("""
            SELECT document_id FROM documents
        """))
        documents = [row[0] for row in result.fetchall()]
        
        print(f"📂 Tìm thấy {len(documents)} documents trong bảng documents\n")
        
        updated = 0
        for doc_id in documents:
            # Đếm chunks cho document này
            result = await session.execute(
                text("""
                    SELECT COUNT(*)
                    FROM langchain_pg_embedding
                    WHERE cmetadata->>'document_id' = :doc_id
                """),
                {"doc_id": doc_id}
            )
            chunk_count = result.scalar()
            
            # Update documents table
            await session.execute(
                text("""
                    UPDATE documents
                    SET total_chunks = :count,
                        updated_at = NOW()
                    WHERE document_id = :doc_id
                """),
                {"count": chunk_count, "doc_id": doc_id}
            )
            
            updated += 1
            if updated % 10 == 0:
                print(f"  ✅ Đã cập nhật {updated}/{len(documents)}...")
        
        await session.commit()
        return updated

updated = await update_chunk_counts()
print(f"\n✅ Đã cập nhật chunk counts cho {updated} documents")

# Verify
async with AsyncSessionLocal() as session:
    result = await session.execute(text("""
        SELECT document_id, total_chunks
        FROM documents
        WHERE total_chunks > 0
        ORDER BY total_chunks DESC
        LIMIT 10
    """))
    
    print(f"\n📊 Top 10 documents theo số lượng chunks:")
    for row in result.fetchall():
        print(f"   {row[0]:40} : {row[1]:4} chunks")
    
    # Check documents without chunks
    result = await session.execute(text("""
        SELECT COUNT(*) FROM documents WHERE total_chunks = 0
    """))
    zero_count = result.scalar()
    
    if zero_count > 0:
        print(f"\n⚠️  {zero_count} documents không có chunks trong vector DB")

await engine.dispose()

🔄 Cập Nhật documents.total_chunks...

📂 Tìm thấy 59 documents trong bảng documents

  ✅ Đã cập nhật 10/59...
  ✅ Đã cập nhật 20/59...
  ✅ Đã cập nhật 30/59...
  ✅ Đã cập nhật 20/59...
  ✅ Đã cập nhật 30/59...
  ✅ Đã cập nhật 40/59...
  ✅ Đã cập nhật 50/59...
  ✅ Đã cập nhật 40/59...
  ✅ Đã cập nhật 50/59...

✅ Đã cập nhật chunk counts cho 59 documents

📊 Top 10 documents theo số lượng chunks:
   FORM-HSYC-HANGHOA-4                      : 1293 chunks
   LUA-HỢP-NHẤT-126-2025-về                 :  617 chunks
   ND-214-4.8.-CP                           :  595 chunks
   TEMPLATE-BC-14A                          :  510 chunks
   LUA-57-2024-QH15                         :  271 chunks
   FORM-7-Mẫu-số-7B-E-HSMT                  :  264 chunks
   LUA-Luat-dau-thau-2023                   :  188 chunks
   FORM-8-Mẫu-số-8B-E-HSMT                  :  166 chunks
   FORM-15-Phu-luc                          :  153 chunks
   FORM-HSYC-TUVAN-01D                      :  134 chunks

⚠️  4 documents không c

## Step 7: Summary Report

**Thời gian:** 30 phút  
**Tạo report về migration**

In [19]:
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from sqlalchemy import text
import os
from datetime import datetime

print("📊 BÁO CÁO TỔNG KẾT MIGRATION")
print("=" * 60)
print(f"Ngày: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")

# Get DATABASE_URL if not already defined
try:
    DATABASE_URL
except NameError:
    DATABASE_URL = os.getenv(
        "DATABASE_URL_ASYNC",
        "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
    )

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def generate_report():
    async with AsyncSessionLocal() as session:
        # Documents table stats
        result = await session.execute(text("""
            SELECT 
                COUNT(*) as total_docs,
                COUNT(*) FILTER (WHERE status = 'active') as active_docs,
                SUM(total_chunks) as total_chunks
            FROM documents
        """))
        row = result.fetchone()
        
        print("📁 Bảng Documents:")
        print(f"   Tổng documents: {row[0]}")
        print(f"   Documents active: {row[1]}")
        print(f"   Tổng chunks (dự kiến): {row[2]}\n")
        
        # Chunks in vector database
        result = await session.execute(text("""
            SELECT 
                COUNT(DISTINCT cmetadata->>'document_id') as unique_docs,
                COUNT(*) as total_chunks,
                COUNT(*) FILTER (WHERE cmetadata->>'source_file' IS NOT NULL) as with_source
            FROM langchain_pg_embedding
        """))
        row = result.fetchone()
        
        print("🗄️  Vector Database:")
        print(f"   Document_ids duy nhất: {row[0]}")
        print(f"   Tổng chunks: {row[1]}")
        print(f"   Có source_file: {row[2]} ({row[2]/row[1]*100:.1f}%)\n")
        
        # By category
        result = await session.execute(text("""
            SELECT category, COUNT(*) as docs, SUM(total_chunks) as chunks
            FROM documents
            GROUP BY category
            ORDER BY docs DESC
        """))
        
        print("📂 Theo Danh Mục:")
        for row in result.fetchall():
            chunks_display = row[2] if row[2] is not None else 0
            print(f"   {row[0]:25} : {row[1]:2} docs, {chunks_display:4} chunks")
        
        # Check for issues
        print(f"\n⚠️  Kiểm Tra Chất Lượng:")
        
        # Documents without chunks
        result = await session.execute(text("""
            SELECT COUNT(*) FROM documents WHERE total_chunks = 0
        """))
        zero_chunks = result.scalar()
        status = "✅" if zero_chunks == 0 else "⚠️"
        print(f"   {status} Documents có 0 chunks: {zero_chunks}")
        
        # Old format chunks
        result = await session.execute(text("""
            SELECT COUNT(*)
            FROM langchain_pg_embedding
            WHERE cmetadata->>'document_id' LIKE '%untitled%'
               OR cmetadata->>'chunk_id' LIKE '%untitled%'
        """))
        old_format = result.scalar()
        status = "✅" if old_format == 0 else "⚠️"
        print(f"   {status} Chunks format cũ còn lại: {old_format}")
        
        # Missing source_file
        result = await session.execute(text("""
            SELECT COUNT(*)
            FROM langchain_pg_embedding
            WHERE cmetadata->>'source_file' IS NULL
               OR cmetadata->>'source_file' = ''
        """))
        no_source = result.scalar()
        status = "✅" if no_source == 0 else "⚠️"
        print(f"   {status} Chunks thiếu source_file: {no_source}")
        
        # Documents mismatch
        docs_in_table = await session.execute(text("SELECT COUNT(*) FROM documents"))
        docs_in_vector = await session.execute(text(
            "SELECT COUNT(DISTINCT cmetadata->>'document_id') FROM langchain_pg_embedding"
        ))
        
        table_count = docs_in_table.scalar()
        vector_count = docs_in_vector.scalar()
        
        print(f"\n📊 So Sánh:")
        print(f"   Documents trong table: {table_count}")
        print(f"   Documents trong vector DB: {vector_count}")
        
        if table_count > vector_count:
            diff = table_count - vector_count
            print(f"   ℹ️  {diff} documents trong table nhưng không có chunks trong vector DB")

await generate_report()

print("\n" + "="*60)
print("✅ MIGRATION HOÀN TẤT")
print("="*60)

await engine.dispose()

📊 BÁO CÁO TỔNG KẾT MIGRATION
Ngày: 2025-11-20 00:26:46

📁 Bảng Documents:
   Tổng documents: 59
   Documents active: 59
   Tổng chunks (dự kiến): 5472

🗄️  Vector Database:
   Document_ids duy nhất: 57
   Tổng chunks: 6242
   Có source_file: 6239 (100.0%)

📂 Theo Danh Mục:
   Hồ sơ mời thầu            : 37 docs, 2873 chunks
   Mẫu báo cáo               : 10 docs,  722 chunks
   Câu hỏi thi               :  4 docs,    0 chunks
   Luật chính                :  4 docs, 1154 chunks
   Thông tư                  :  2 docs,  123 chunks
   Nghị định                 :  1 docs,  595 chunks
   Quyết định                :  1 docs,    5 chunks

⚠️  Kiểm Tra Chất Lượng:
   ⚠️ Documents có 0 chunks: 4
   ⚠️ Chunks format cũ còn lại: 770
   ⚠️ Chunks thiếu source_file: 770

📊 So Sánh:
   Documents trong table: 59
   Documents trong vector DB: 57
   ℹ️  2 documents trong table nhưng không có chunks trong vector DB

✅ MIGRATION HOÀN TẤT


---

## 🎯 Next Steps

Sau khi migration hoàn tất, cần update:

### 1. API Endpoints
- [ ] Update catalog endpoint để query từ documents table
- [ ] Update toggle endpoint để support document_id
- [ ] Add filtering by category, document_type

### 2. Retrieval Pipeline
- [ ] Update retriever để filter by document status
- [ ] Add document_id filter support
- [ ] Test retrieval với new document_ids

### 3. Testing
- [ ] Test catalog API: Should return 70 documents
- [ ] Test toggle: Deactivate specific document
- [ ] Test retrieval: Should exclude inactive documents
- [ ] Performance testing

### 4. Documentation
- [ ] Update API documentation
- [ ] Document new document_id format
- [ ] Create migration runbook

---

## 🗑️ Optional: Cleanup Old Collections

**Thời gian:** 5 phút  
**Mục đích:** Xóa langchain_pg_collection cũ sau khi verify thành công

⚠️ **CHỈ CHẠY SAU KHI:**
- ✅ Step 5 verify: 100% chunks có document_id mới
- ✅ Step 6 complete: documents.total_chunks đã sync
- ✅ Step 7 report: Không có lỗi nào

**Lý do cần cleanup:**
- `langchain_pg_collection` có 5 collections cũ (không dùng nữa)
- `langchain_pg_embedding.collection_id` vẫn trỏ về collections cũ
- Sau migration, chúng ta chỉ dùng `cmetadata->>'document_id'` mới

In [ ]:
print("🗑️  CLEANUP OLD COLLECTIONS")
print("="*60)

# Get DATABASE_URL if not already defined
try:
    DATABASE_URL
except NameError:
    DATABASE_URL = os.getenv(
        "DATABASE_URL_ASYNC",
        "postgresql+asyncpg://sakana:sakana123@localhost/rag_bidding_v2"
    )

engine = create_async_engine(DATABASE_URL, echo=False)
AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)

async def cleanup_collections():
    async with AsyncSessionLocal() as session:
        # 1. Check current collections
        result = await session.execute(text("""
            SELECT uuid, name, cmetadata 
            FROM langchain_pg_collection
            ORDER BY name
        """))
        collections = result.fetchall()
        
        print(f"\n📋 Current collections in langchain_pg_collection:")
        for col in collections:
            print(f"   • {col[1]} (UUID: {col[0]})")
        
        if not collections:
            print("   ✅ No collections found - already cleaned!")
            return
        
        # 2. Check if any chunks still using collection_id
        result = await session.execute(text("""
            SELECT COUNT(*) 
            FROM langchain_pg_embedding 
            WHERE collection_id IS NOT NULL
        """))
        chunks_with_collection = result.scalar()
        
        print(f"\n📊 Chunks still linked to collections: {chunks_with_collection}")
        
        # 3. Option 1: Unlink chunks from collections (RECOMMENDED)
        print(f"\n🔧 Option 1: Unlink chunks (set collection_id = NULL)")
        print("   This keeps collections but removes the link.")
        print("   Safer option - can still rollback if needed.\n")
        
        user_input = input("   Execute Option 1? (yes/no): ").strip().lower()
        
        if user_input == 'yes':
            await session.execute(text("""
                UPDATE langchain_pg_embedding
                SET collection_id = NULL
            """))
            await session.commit()
            print("   ✅ All chunks unlinked from collections")
        
        # 4. Option 2: Delete collections (PERMANENT)
        print(f"\n🗑️  Option 2: DELETE collections permanently")
        print("   ⚠️  WARNING: This is IRREVERSIBLE!")
        print("   ⚠️  Only do this after 100% verification!\n")
        
        user_input = input("   Execute Option 2? (yes/no): ").strip().lower()
        
        if user_input == 'yes':
            # First, unlink all chunks
            await session.execute(text("""
                UPDATE langchain_pg_embedding
                SET collection_id = NULL
            """))
            
            # Then delete collections
            await session.execute(text("""
                DELETE FROM langchain_pg_collection
            """))
            
            await session.commit()
            print("   ✅ Collections deleted permanently")
            print("   ✅ All chunks unlinked")
        
        # 5. Verify cleanup
        result = await session.execute(text("""
            SELECT COUNT(*) FROM langchain_pg_collection
        """))
        remaining_collections = result.scalar()
        
        result = await session.execute(text("""
            SELECT COUNT(*) 
            FROM langchain_pg_embedding 
            WHERE collection_id IS NOT NULL
        """))
        remaining_links = result.scalar()
        
        print(f"\n{'='*60}")
        print(f"📊 After cleanup:")
        print(f"   Collections remaining: {remaining_collections}")
        print(f"   Chunks with collection_id: {remaining_links}")
        print(f"{'='*60}")

await cleanup_collections()
await engine.dispose()